itou/www/signup/forms.py

```diff
@@ -91,50 +95,159 @@ def save(self, request):
         return user
 
 
-class SiaeSignupForm(FullnameFormMixin, SignupForm):
+class SelectSiaeForm(forms.ModelForm):
+    """
+    First of two forms of siae signup process.
+    This first form allows the user to select which siae will be joined.
+    """
+
+    class Meta:
+        model = Siae
+        fields = ["kind", "siret", "email"]
+
+    kind = forms.ChoiceField(
+        label=_("Type de votre SIAE"),
+        choices=BLANK_CHOICE + Siae.KIND_CHOICES,
+        required=True,
+    )
 
     siret = forms.CharField(
         label=_("Numéro SIRET de votre SIAE"),
+        min_length=14,
         max_length=14,
         validators=[validate_siret],
-        required=True,
         strip=True,
-        help_text=_("Saisissez 14 chiffres."),
+        help_text=_(
+            "Saisissez 14 chiffres. Doit si possible être le SIRET connu de l'ASP."
+        ),
+        required=False,
     )
 
-    def clean_siret(self):
-        siret = self.cleaned_data["siret"]
+    email = forms.EmailField(
+        label=_("E-mail"),
+        help_text=_("Doit si possible être l'adresse e-mail connue de l'ASP."),
+        required=False,
+    )
 
-        siret_exists_in_db = Siae.active_objects.filter(siret=siret).exists()
-        if siret_exists_in_db:
-            return siret
+    def save(self, request, commit=True):
+        raise RuntimeError("SelectSiaeForm.save() should never be called.")
+
+    # pylint: disable=inconsistent-return-statements
+    def clean(self):
+        kind = self.cleaned_data["kind"]
+        siret = self.cleaned_data["siret"]
+        email = self.cleaned_data["email"]
+
+        if not siret and not email:
```

Comment:
Je pense toujours que c'est un `or` et pas un `and` qu'il te faut ici :)

Reply:
Aussi je pense qu'il y a un pattern qui te permettrai d'avoir un code plus simple à lire et avec moins de requêtes. Il s'agit de construire le `lookup` de ta requête et de l'exécuter une seule fois. À titre d'info perso j'aurais fait quelque chose comme ça :

```python
kwargs = {
    "kind": kind,
}
if siret:
    kwargs["siret"] = siret
if email:
    kwargs["auth_email"] = email

siaes = Siae.active_objects.filter(**kwargs)

if not siaes.exists():
    raise("No result")

if siaes.count() > 1:
    raise("Multiple results, which SIRET?")

return siaes.first()
```

Reply:
> Je pense toujours que c'est un or et pas un and qu'il te faut ici

ben non justement, sur ta suggestion, il n'y a plus besoin que de remplir 1 de ces 2 champs, du coup si l'un des 2 est vide on laisse passer

l'opérateur `not` a priorité sur l'opérateur `and` hein...

je vais remplacer `not siret and not email` par l'équivalent `not(siret or email)` pour moins porter à confusion

Reply:
Ton bout de code est correct seulement pour les cas où on a soit un email soit un siret.

Dans le cas où les 2 sont remplis, on échappe pas à faire 2 requêtes. Il faut d'abord voir si une siae matche le siret, et uniquement dans le cas contraire, faire une second tentative pour voir si une siae matche l'email.

Je ne peux pas non plus fusionner en un unique requête du genre `filter(siret=siret OR email=email)` car elle ne me permettrait pas le contrôle fin des edge cases dont j'ai besoin pour retourner les bons messages d'erreur.

Reply:
> l'opérateur not a priorité sur l'opérateur and hein...

Ah oui tu as raison j'ai lu trop vite

Reply:
> Ton bout de code est correct seulement pour les cas où on a soit un email soit un siret.

> Dans le cas où les 2 sont remplis, on échappe pas à faire 2 requêtes. Il faut d'abord voir si une siae matche le siret, et uniquement dans le cas contraire, faire une seconde tentative pour voir si une siae matche l'email.

C'est justement l'inverse que je dis.

Tu peux faire les contrôles fins sans faire de requête supplémentaire puisque tu sais que tu as une unicité SIRET-Type, le seul cas dans lequel tu peux avoir plusieurs résultats c'est avec un email.

Du coup une seule query te suffit et tu sors du form avec un objet SIAE qui t'évite encore des requêtes dans ta view et un try dans un try : https://github.com/betagouv/itou/pull/40/files#diff-2f710d61d118057f0bad8f18dd2fee91R56

Reply:
T'as complètement raison en fait. Je vais creuser pour simplifier cela.